dashboard/test/controllers/openai_chat_controller_test.rb

```diff
@@ -0,0 +1,30 @@
+require 'test_helper'
+
+class OpenaiChatControllerTest < ActionController::TestCase
+  setup do
+    response = Net::HTTPResponse.new(nil, '200', nil)
+    OpenaiChatHelper.stubs(:request_chat_completion).returns(response)
+    OpenaiChatHelper.stubs(:get_chat_completion_response_message).returns({status: 200, json: {}})
+  end
+
+  # User without ai_chat_access is unable to access the chat completion endpoint
+  test_user_gets_response_for :chat_completion,
+  user: :levelbuilder,
+  method: :post,
+  params: {messages: [{role: "user", content: "Say this is a test!"}]},
+  response: :forbidden
+
+  # With ai_chat_access, a post request without a messages param returns a bad request
+  test_user_gets_response_for :chat_completion,
+  user: :ai_chat_access,
+  method: :post,
+  params: {},
+  response: :bad_request
+
+  # With ai_chat_access, a post request with a messages param returns a success
+  test_user_gets_response_for :chat_completion,
+  user: :ai_chat_access,
+  method: :post,
+  params: {messages: [{role: "user", content: "Say this is a test!"}]},
+  response: :success
+end
```

Comment:
nice test cases! based on my experience so far, you are likely to hit a rate limit error (429) from OpenAI at some point. How about a test for the case where OpenAI returns status 429, and you verify that that status code gets passed through?

FWIW it looks like your code is set up to handle this already, so this is mostly to cover regressions.

Reply:
Thanks! I have a separate [PR](https://github.com/code-dot-org/code-dot-org/pull/52976) which adds throttle logic and add a test for status 429. If you have time, would love for you to take a look - thanks!